### PR TITLE
pipeline: decouple resources from pipeline creation

### PR DESCRIPTION
**Pull request by mbouron-gpsw**
Based on https://github.com/gopro/gopro-lib-node.gl/pull/92, rebased on master.

### CODE REVIEW

**Comment by mbouron-gpsw** [comment]
Branch updated:
- renamed bind_resources to set_resources
- bunch of cosmetics fixes
- added checks in ngli_pipeline_gl_set_resources() to make sures the resource array has the same size as the the description array.

File: libnodegl/darray.h [hunk 1]
```diff
@@ -38,6 +38,7 @@ void *ngli_darray_push(struct darray *darray, const void *element);
 void *ngli_darray_pop(struct darray *darray);
 void *ngli_darray_tail(const struct darray *darray);
 void *ngli_darray_get(const struct darray *darray, int index);
+int ngli_darray_set(const struct darray *darray, int index, const void *element);
```
Comment (cboesch-gpsw) [inline libnodegl/darray.h:41]: Please keep this commit out until it is needed; I'm not a fan of uncovered code.

File: libnodegl/pgcraft.c [hunk 2]
```diff
@@ -924,10 +935,14 @@ static int probe_pipeline_elems(struct pgcraft *s)
...
+    if ((ret = filter_pipeline_elems(s, probe_pipeline_uniform,   uniforms_info,   &s->pipeline_info.desc.uniforms,   &s->pipeline_info.data.uniforms,
+                                     &s->filtered_pipeline_info.desc.uniforms, &s->filtered_pipeline_info.data.uniforms))   < 0 ||
+        (ret = filter_pipeline_elems(s, probe_pipeline_buffer,    buffers_info,    &s->pipeline_info.desc.buffers,    &s->pipeline_info.data.buffers,
+                                     &s->filtered_pipeline_info.desc.buffers,    &s->filtered_pipeline_info.data.buffers))    < 0 ||
+        (ret = filter_pipeline_elems(s, probe_pipeline_texture,   uniforms_info,   &s->pipeline_info.desc.textures,   &s->pipeline_info.data.textures,
+                                     &s->filtered_pipeline_info.desc.textures,    &s->filtered_pipeline_info.data.textures))   < 0 ||
+        (ret = filter_pipeline_elems(s, probe_pipeline_attribute, attributes_info, &s->pipeline_info.desc.attributes, &s->pipeline_info.data.attributes,
+                                     &s->filtered_pipeline_info.desc.attributes,    &s->filtered_pipeline_info.data.attributes)) < 0)
```
Comment (cboesch-gpsw) [inline libnodegl/pgcraft.c:945]: can you make the indent/align more pleasing to the eye?

File: libnodegl/pgcraft.h [hunk 3]
```diff
@@ -149,22 +149,30 @@ enum {
...
+        struct darray uniforms;   // uniform_desc
+        struct darray textures;   // texture_desc
+        struct darray buffers;    // buffer_desc
+        struct darray attributes; // attribute_desc
+    } desc;
+    struct {
+        struct darray uniforms;   // uniform data pointer
+        struct darray textures;   // texture
+        struct darray buffers;    // buffer
+        struct darray attributes; // attribute
```
Comment (cboesch-gpsw) [inline libnodegl/pgcraft.h:163]: this is incorrect: those are pointers as well

File: libnodegl/pipeline.h [hunk 4]
```diff
@@ -100,11 +107,13 @@ struct pipeline {
     struct darray texture_descs;
     struct darray buffer_descs;
     struct darray attribute_descs;
+
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline.h:110]: unrelated

File: libnodegl/pipeline_gl.c [hunk 5]
```diff
@@ -323,21 +313,20 @@ static void set_vertex_attribs(const struct pipeline *s, struct glcontext *gl)
 {
     const struct attribute_desc *descs = ngli_darray_data(&s->attribute_descs);
     for (int i = 0; i < ngli_darray_count(&s->attribute_descs); i++) {
-        const struct attribute_desc *desc = &descs[i];
-        const struct pipeline_attribute *attribute = &desc->attribute;
-        const struct buffer *buffer = attribute->buffer;
+        const struct attribute_desc *attr_desc = &descs[i];
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_gl.c:316]: Unless I'm mistaken, `struct {uniform,texture,buffer,attribute}_desc` are going to be renamed to something else, because they also contain the data. So If you're renaming that variable `desc`, please don't rename it to something that includes `desc`, because it will remain as such after the rename of the structs. Maybe `attr` is more appropriate (`attr->desc.xxx`, `attr->data.xxx`).

File: libnodegl/pipeline_gl.c [hunk 6]
```diff
@@ -346,9 +335,8 @@ static void reset_vertex_attribs(const struct pipeline *s, struct glcontext *gl)
 {
     const struct attribute_desc *descs = ngli_darray_data(&s->attribute_descs);
     for (int i = 0; i < ngli_darray_count(&s->attribute_descs); i++) {
-        const struct attribute_desc *desc = &descs[i];
-        const struct pipeline_attribute *attribute = &desc->attribute;
-        const GLuint location = attribute->location;
+        const struct attribute_desc *attr_desc = &descs[i];
+        const GLuint location = attr_desc->desc.location;
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_gl.c:339]: same remark here

File: libnodegl/pipeline_gl.c [hunk 7]
```diff
@@ -391,6 +377,18 @@ static GLenum get_gl_indices_type(int indices_format)
     return gl_indices_type_map[indices_format];
 }
 
+static void init_vertex_attribs(const struct pipeline *s, struct glcontext *gl)
+{
+    const struct attribute_desc *descs = ngli_darray_data(&s->attribute_descs);
+    for (int i = 0; i < ngli_darray_count(&s->attribute_descs); i++) {
+        const struct attribute_desc *attr_desc = &descs[i];
+        const GLuint location = attr_desc->desc.location;
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_gl.c:385]: and here

File: libnodegl/pipeline_gl.c [hunk 8]
```diff
@@ -479,6 +477,40 @@ int ngli_pipeline_gl_init(struct pipeline *s, const struct pipeline_params *para
...
+    ngli_assert(ngli_darray_count(&s->texture_descs) == data_params->nb_textures);
+    for (int i = 0; i < data_params->nb_textures; i++) {
+        int ret = ngli_pipeline_gl_update_texture(s, i, data_params->textures[i]);
+        if (ret < 0)
+            return ret;
+    }
+
+    ngli_assert(ngli_darray_count(&s->uniform_descs) == data_params->nb_uniforms);
+    for (int i = 0; i < data_params->nb_uniforms; i++) {
+        struct uniform_desc *uniform_desc = ngli_darray_get(&s->uniform_descs, i);
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_gl.c:505]: `uniform_desc` → `uniform`

File: libnodegl/pipeline_gl.c [hunk 9]
```diff
@@ -494,28 +526,26 @@ int ngli_pipeline_gl_update_attribute(struct pipeline *s, int index, struct buff
...
     if (!buffer)
         return 0;
 
     if (gl->features & NGLI_FEATURE_VERTEX_ARRAY_OBJECT) {
-        const GLuint location = attribute->location;
-        const GLuint size = ngli_format_get_nb_comp(attribute->format);
-        const GLint stride = attribute->stride;
+        const GLuint location = attr_desc->desc.location;
+        const GLuint size = ngli_format_get_nb_comp(attr_desc->desc.format);
+        const GLint stride = attr_desc->desc.stride;
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_gl.c:543]: This one is unfortunate because the `attr_desc` already exists, so it should probably be left as is in this instance.

I'd appreciate if a naming cleanup was following this PR.

File: libnodegl/pipeline_gl.c [hunk 10]
```diff
@@ -494,28 +526,26 @@ int ngli_pipeline_gl_update_attribute(struct pipeline *s, int index, struct buff
...
         struct buffer_gl *buffer_gl = (struct buffer_gl *)buffer;
         ngli_glBindVertexArray(gl, s_priv->vao_id);
         ngli_glBindBuffer(gl, GL_ARRAY_BUFFER, buffer_gl->id);
-        ngli_glVertexAttribPointer(gl, location, size, GL_FLOAT, GL_FALSE, stride, (void*)(uintptr_t)(attribute->offset));
+        ngli_glVertexAttribPointer(gl, location, size, GL_FLOAT, GL_FALSE, stride, (void*)(uintptr_t)(attr_desc->desc.offset));
     }
-
```
Comment (cboesch-gpsw) [inline libnodegl/pipeline_gl.c:518]: unrelated

**Review by cboesch-gpsw — COMMENTED** [review]
I only have cosmetics comments remaining on this PR, rest LGTM

File: libnodegl/pipeline_gl.c (hunk 10 above)
Comment (mbouron-gpsw) [reply libnodegl/pipeline_gl.c:518]: Fixed.

File: libnodegl/pipeline.h (hunk 4 above)
Comment (mbouron-gpsw) [reply libnodegl/pipeline.h]: Fixed.

File: libnodegl/darray.h (hunk 1 above)
Comment (mbouron-gpsw) [reply libnodegl/darray.h:41]: Done.

File: libnodegl/pgcraft.h (hunk 3 above)
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.h:163]: Fixed.

File: libnodegl/pgcraft.c (hunk 2 above)
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.c]: Done.

File: libnodegl/pipeline_gl.c (hunk 5 above)
Comment (mbouron-gpsw) [reply libnodegl/pipeline_gl.c]: I added a commit doing the rename on top of the branch. It renames the suffix _desc with _binding for all structures, fields and variables in pipeline_gl. It also expand the name attr to attribute so it is consistent with the other declarations (ie: buffer_binding, texture_binding, attribute_binding).

**Review by cboesch-gpsw — APPROVED** [review]
